app/Http/Middleware/VersionSwitcher.php

```diff
@@ -27,28 +26,17 @@ public function handle($request, $next)
 
         $route = $request->route();
 
-        $request->setRouteResolver(function () use ($version, $route) {
-
-            return array_map(function ($routeComponent) use ($version) {
-                // Only handle routes that has a controller defined.
-                if ($uses = $routeComponent['uses'] ?? null) {
-                    if (preg_match('/^([^@]+)@(.*)$/', $uses, $m)) {
-                        list(, $classFrom, $method) = $m;
-
-                        // Squeeze in a \v[version]\ before the controller class name.
-                        $replacement = sprintf('$1v%d\\\\$2', $version);
-                        $classTo = preg_replace('/(.*\\\\)([\w]+)$/', $replacement, $classFrom);
-
-                        // If we have a controller candidate use that.
-                        if (method_exists($classTo, $method)) {
-                            $routeComponent['uses'] = "$classTo@$method";
-                        }
-                    }
+        foreach ($route as $routeComponent) {
+            // Only handle routes that has a controller defined.
+            if ($uses = $routeComponent['uses'] ?? null) {
+                [$controllerFrom, $method] = explode('@', $uses);
+                $replacement = sprintf('$1v%d\\\\$2', $version);
+                $controllerTo = preg_replace('/(.*\\\\)([\w]+)$/', $replacement, $controllerFrom);
+                if (method_exists($controllerTo, $method)) {
```

Comment:
Just do a `class_exists' and let Lumen deal with missing methods.

Reply:
I didn't mean to imply it was done without thought.

But neither was my suggestion. If v2 doesn't implement a method, do we *really* want to use the v1 implementation? The client has explicitly asked for v2, and if the developer has messed up, we'd rather have it error, won't we? And if a path has been removed in the new version, we want it do disappear.